### PR TITLE
Using column_sampler for optimization of ColWiseBuildHist

### PR DESCRIPTION
**Pull request by razdoburdin**
Hi,
this PR is a part of https://github.com/dmlc/xgboost/pull/7192. 
It continues the optimization of BuildHistKernel, which was started at https://github.com/dmlc/xgboost/pull/8233.
In https://github.com/dmlc/xgboost/pull/8233 I removed the optimization related to column_sampler to simplify the reviewing process. Here I return it. 

The methodology of the benchmarking is the same as in https://github.com/dmlc/xgboost/pull/8218. The only dataset affected by optimization is `santander`. Training speed for `santander `increases twice after applying this optimization. You can find the exact numbers at https://github.com/dmlc/xgboost/pull/8233.

I am looking forward for your review and comments!

### CODE REVIEW

**Comment by hcho3** [comment]
Please ignore the failing MacOS tests from GitHub Actions. We're working to fix it.

**Comment by razdoburdin** [comment]
Hi @trivialfis ,
what is your opinion about this optimization?

**Comment by trivialfis** [comment]
Apologies for the slow response, I will get to this tomorrow.

**Comment by razdoburdin** [comment]
> Apologies for the slow response, I will get to this tomorrow.

Hi,
have you any decision about this PR?

File: src/common/hist_util.cc [hunk 1]
```diff
@@ -284,13 +300,14 @@ void ColsWiseBuildHistKernel(const std::vector<GradientPair> &gpair,
   };
 
   const size_t n_features = gmat.cut.Ptrs().size() - 1;
-  const size_t n_columns = n_features;
+  const size_t n_columns = kColumnSampling ? fids.size() : n_features;
```
Comment (trivialfis) [inline src/common/hist_util.cc:303]: Is `fids` empty if there's no sampling?

File: src/tree/hist/histogram.h [hunk 2]
```diff
@@ -76,6 +84,20 @@ class HistogramBuilder {
       buffer_.Reset(this->n_threads_, n_nodes, space, target_hists);
     }
 
+    constexpr float kColsampleTh = 0.1;
+    bool column_sampling = (column_sampler_ != nullptr) &&
```
Comment (trivialfis) [inline src/tree/hist/histogram.h:88]: When column sampler is nullptr?

File: src/tree/hist/histogram.h [hunk 3]
```diff
@@ -76,6 +84,20 @@ class HistogramBuilder {
       buffer_.Reset(this->n_threads_, n_nodes, space, target_hists);
     }
 
+    constexpr float kColsampleTh = 0.1;
```
Comment (trivialfis) [inline src/tree/hist/histogram.h:87]: why 0.1?

File: src/tree/hist/histogram.h [hunk 4]
```diff
@@ -76,6 +84,20 @@ class HistogramBuilder {
...
+    bool column_sampling = (column_sampler_ != nullptr) &&
+                           (train_param_.colsample_bytree < kColsampleTh ||
+                            train_param_.colsample_bylevel < kColsampleTh);
```
Comment (trivialfis) [inline src/tree/hist/histogram.h:90]: how about bynode? Is it used?

**Review by trivialfis — COMMENTED** [review]
Apologies for the delay. May I ask the specific objective of this PR? Are you trying to optimize the case to build for less columns when sampling is enabled?

I'm concerned with optimizations that don't generalize and requires edge cases. I haven't been able to track which test can confirm the correctness of the code.

File: src/common/hist_util.cc (hunk 1 above)
Comment (razdoburdin) [reply src/common/hist_util.cc:303]: Fids is empty is case of condition [here](https://github.com/dmlc/xgboost/pull/8319/files#diff-8217fa4498a3341ec017b9661c7cc75123cc88fb058c865db9a7f71558a01cb0R91) is false.

File: src/tree/hist/histogram.h (hunk 2 above)
Comment (razdoburdin) [reply src/tree/hist/histogram.h:88]: I set it to nullptr in [tests ](https://github.com/dmlc/xgboost/pull/8319/files#diff-4bdebd351f1f1f377c74e6764eaf56ff9c6e1b5f9a123ab51665003f164b56d2R49)of the histogram builder without using this optimization.

File: src/tree/hist/histogram.h (hunk 3 above)
Comment (razdoburdin) [reply src/tree/hist/histogram.h:87]: It is an ad-hoc threshold value.

File: src/tree/hist/histogram.h (hunk 4 above)
Comment (razdoburdin) [reply src/tree/hist/histogram.h:90]: Not now, maybe later, one can investigate these options more deeply.

**Comment by razdoburdin** [comment]
> Apologies for the delay. May I ask the specific objective of this PR? Are you trying to optimize the case to build for less columns when sampling is enabled?
yes, that was the idea

> I'm concerned with optimizations that don't generalize and requires edge cases. I haven't been able to track which test can confirm the correctness of the code.
It is reasonable. I added a [subcase](https://github.com/dmlc/xgboost/pull/8319/files#diff-42fecfa97f179a4084d1ec4e4a86e845882f64d6acf1d46d03b480010b31dee5R47) for `TestEvaluateSplits ` that forces using the code with` kColumnSampling = true`. The idea is to fill `fids `vector manually. It is enough for pushing the execution to the branch with `kColumnSampling = true`.

**Comment by trivialfis** [comment]
Let me try to look deeper into this PR. The optimization is definitely important but I would like to keep lesser conditional branches than it currently has.
